### DISCUSSION TITLE
How to correctly pre-install and load packages in a modified console.html?

### DISCUSSION THREAD

**Discussion by johann-petrak** [pyodide/pyodide#2703] [Q&A]
I have obtained a pyodide release and serve it from my own server. That release contains console.html for showing the Python REPL in a web page.

What would be the correct way to run custom code in that REPL before the prompt is available to the user? 
Ideally the REPL would show, some printed message would get shown to the user in the REPL window, micropip would install a bunch of packages, the packages would get imported, some additional messages would get shown to the user and then the REPL would be ready for interaction. Where/how in the console.html to add this correctly?

**Comment by ryanking13** [comment]
I think sympy team is doing some similar thing (https://sympy-beta.vercel.app/terminal). @eagleoflqj Are you interested in sharing your experience with this?

**Comment by eagleoflqj** [comment]
If SymPy Beta's behavior is what you want, then you may reference this piece of code after calling `jQuery(el).terminal(...)`
https://github.com/eagleoflqj/sympy_beta/blob/8a298b64808c933163bb3d1d8fd7e9da679b73f1/src/views/TerminalView.vue#L102-L105
I use a variable `loaded` to indicate whether pyodide is ready. So when it changes to `true`, Vue lets the skeleton disappear.

**Comment by jobovy** [comment]
Another option might be to run a [`jupyterlite` REPL](https://jupyterlite.readthedocs.io/en/latest/quickstart/embed-repl.html#embed-a-live-repl-on-a-website), which has a `code` URL parameter that you can use to pre-run code. This code is shown to the user. If you want to run code before that that isn't shown to the user, you could use my fork of `jupyterlite` that adds a `prerun` parameter with code that will be run without showing the user (see https://github.com/jupyterlite/jupyterlite/issues/508#issuecomment-1051479045). The maintainers over at `jupyterlite` don't want to merge this, but I think it's quite useful to hide `pyodide/jupyterlite` specific code that is confusing to a regular Python user.

I have an example of implementing this at https://github.com/jobovy/galpy-repl, where I also have a workflow that will take a Python file and create the `prerun=` parameter for you so you can prerun complex code without having to URL-encode it yourself (you can also run that to get the `code=` parameter. You should be able to relatively easily adjust the workflow to build your own `jupyterlite` distribution there.

**Reply by loosak** [reply]
Simply  load micropip and add you custom  into  `function main()` of [console.html](https://github.com/pyodide/pyodide/blob/88313f76d5a4837248e1cea248ad66b46f44e399/src/templates/console.html) this way

```
...
        await pyodide.loadPackage("micropip");  //  load micropip package
        await pyodide.runPythonAsync(                
          `
            import sys
            from pyodide import to_js
            from pyodide.console import PyodideConsole, repr_shorten, BANNER
            import __main__
            BANNER = "Welcome to the Pyodide terminal emulator 🐍\\n" + BANNER
            pyconsole = PyodideConsole(__main__.__dict__)
            import builtins
            async def await_fut(fut):
              res = await fut
              if res is not None:
                builtins._ = res
              return to_js([res], depth=1)
            def clear_console():
              pyconsole.buffer = []
            # Add some modules 
            import micropip
            await micropip.install(['regex', 'pandas', 'numpy', 'matplotlib'])
            # other custom code ...
            BANNER += "\\nregex, pandas, numpy, matplotlib has been installed!\\n"
        `,
...
````

**Reply by johann-petrak** [reply]
Thank you @loosak that allows me to pre-install packages but it does not allow me to pre-import anything: if I add something like "import numpy as np" after the micropip.install line, np is not visible in the console. 
Is it correct to use something like `pyconsole.push("import numpy as np")` within the runPythonAsync function to do that or is there a better/proper way to do that?

**Reply by loosak** [reply]
Sorry, but I don`t understand what do you mean by:
> "import numpy as np" after the micropip.install line, np is not visible in the console

import numpy  should work after instalation by micropip,  but take into considerations, that micropip is asynchronous and should be awaited before import, as can you see in my snipped above.
You can also install python packages from js  by `pyodide.loadPackage()` (as I used for micropip) and  in this case you don`t  need use micropip in your python code. In that case you have fully compatible python code with os version (if your modules are supported by pyodide).

**Comment by johann-petrak** [comment]
When I just add the line `import numpy as np` after the `await micropip.install(['regex', 'pandas', 'numpy', 'matplotlib'])` then the numpy module gets installed in that context and is visible in subsequent lines. 
However the module is NOT visible to the user in the console! Numpy is installed and can be imported by the user, but if it has been imported there, then the user will not find it imported.
So I tried to instead run `pyconsole.push("import numpy as np")` instead of running `import numpy as np` and with this numpy is indeed visible as np to the user in the console.

I was wondering if this is indeed the correct way to do this.

**Reply by loosak** [reply]
OK. finally I got it. Python code in console.py is interpreted under PyodideConsole object local namespace, not pyodide's namespace. Therefore, after you import all modules (define vars etc), put pyodide locals to PyodideConsole namespace simply by:

 `pyconsole = PyodideConsole(locals())`

Here is my working snippets of changes in console.html
```
...
        await pyodide.loadPackage("micropip");  // load micropip package
        await pyodide.runPythonAsync(           // change runPythonAsync
          `
            import sys
            from pyodide import to_js
            from pyodide.console import PyodideConsole, repr_shorten, BANNER
            BANNER = "Welcome to the Pyodide terminal emulator 🐍\\n" + BANNER
            # Add some modules
            import micropip
            await micropip.install(['pandas', 'numpy'])
            import numpy as np
            import pandas as pd
            import __main__
            BANNER += "\\npandas, numpy has been installed and imported!\\n"
            pyconsole = PyodideConsole(locals())
...
```
Example of prepared env in the app:

```
Welcome to the Pyodide terminal emulator 🐍
Python 3.10.2 (main, Apr  9 2022 20:52:01) on WebAssembly VM
Type "help", "copyright", "credits" or "license" for more information.
pandas, numpy has been installed and imported!
>>> pd.DataFrame({'a': np.random.rand(5)})
          a
0  0.149881
1  0.799378
2  0.469033
3  0.300900
4  0.580101
>>> 
 
```
